### PR TITLE
Timestamp in recipient

### PR DESCRIPTION
**Pull request by fboucquez**
Fixes https://github.com/symbol/catapult-rest/issues/636

It adds timestamp next to the id in transaction statements.

image

This implementation runs a second query looking for blocks for the given block heights. It does not use aggregate queries. I tried but failed to create an aggregate query between the `transactionStatements` and `blocks` collections.

If a mongo aggregate query is desired, I can try again to make it work. Note that the aggregate is between nested properties, not top-level properties.

### CODE REVIEW

**Comment by fboucquez** [comment]
Once this is approved, the solution can be replicated for https://github.com/symbol/catapult-rest/issues/637

File: catapult-sdk/test/plugins/receipts_spec.js
Comment (Jaguar0625) [inline catapult-sdk/test/plugins/receipts_spec.js:63]: you might want to update this check

File: rest/test/plugins/receipts/ReceiptsDb_spec.js
Comment (Jaguar0625) [inline rest/test/plugins/receipts/ReceiptsDb_spec.js:193]: why did you need to change these?

File: rest/test/db/utils/dbTestUtils.js
Comment (Jaguar0625) [inline rest/test/db/utils/dbTestUtils.js:272]: i think insertEntities would be more consistent with other naming in this file

File: rest/src/db/CatapultDb.js
Comment (Jaguar0625) [inline rest/src/db/CatapultDb.js:235]: why this projection?

File: rest/test/db/CatapultDb_spec.js
Comment (Jaguar0625) [inline rest/test/db/CatapultDb_spec.js:447]: consider grouping these in own describe

File: rest/src/db/CatapultDb.js
Comment (Jaguar0625) [reply rest/src/db/CatapultDb.js:235]: don't you only need timestamp?

File: rest/test/db/CatapultDb_spec.js
Comment (Jaguar0625) [reply rest/test/db/CatapultDb_spec.js:447]: consider swapping order of tests so they go from simpler to more complex

File: rest/src/db/CatapultDb.js
Comment (fboucquez) [reply rest/src/db/CatapultDb.js:235]: > why this projection?

To keep it consistent with the singular `blockAtHeight` 
https://github.com/symbol/catapult-rest/blob/e730a5ea4d8c25fc4aedf534962de4040eef3c83/rest/src/db/CatapultDb.js#L220. 

> don't you only need timestamp?

In this case yes, for #637 fix I would need the feeMultipler. 

Having said that, I'll add the custom projection.

File: rest/test/plugins/receipts/ReceiptsDb_spec.js
Comment (fboucquez) [reply rest/test/plugins/receipts/ReceiptsDb_spec.js:193]: I was testing the unique ids, it's tested elsewhere so rolling this back

File: catapult-sdk/test/plugins/receipts_spec.js
Comment (fboucquez) [reply catapult-sdk/test/plugins/receipts_spec.js:63]: Ok, I see, thanks. Would we want to add timestamp to `addressResolutionStaments` and `mosaicResolutionStatements`? 

To these endpoints?

http://ngl-dual-101.testnet.symboldev.network:3000/statements/resolutions/address
http://ngl-dual-101.testnet.symboldev.network:3000/statements/resolutions/mosaic

Or separate transactionStements from the other 2 is fine?

**Comment by fboucquez** [comment]
@Jaguar0625 , do you like the location of the timestamp next to id? I'm not sure where is the best place, top-level timestamp, inside statement, in a meta object/field, or in a (partial) block object/field.

**Comment by Jaguar0625** [comment]
> @Jaguar0625 , do you like the location of the timestamp next to id? I'm not sure where is the best place, top-level timestamp, inside statement, in a meta object/field, or in a (partial) block object/field.

i think in meta would probably be preferred. @gimre-xymcity do you have any opinion?

File: rest/test/db/CatapultDb_spec.js
Comment (fboucquez) [reply rest/test/db/CatapultDb_spec.js:464]: Added describe, a new projection unit test and resorted tests.

File: rest/src/db/CatapultDb.js
Comment (fboucquez) [reply rest/src/db/CatapultDb.js]: Added custom projection + unit tests

**Comment by fboucquez** [comment]
Remaining questions:
1 - Where to put the timestamp in the response?
2 - Should we add the timestamp to mosaic and address statements for consistency?

**Comment by gimre-xymcity** [comment]
> > @Jaguar0625 , do you like the location of the timestamp next to id? I'm not sure where is the best place, top-level timestamp, inside statement, in a meta object/field, or in a (partial) block object/field.
> 
> i think in meta would probably be preferred. @gimre-xymcity do you have any opinion?

meta is probably most appropriate

**Comment by fboucquez** [comment]
@Jaguar0625 , I've moved the timestamp to the meta field and I have added the meta/timestamp to address and mosaic statements. 

image

File: catapult-sdk/test/plugins/receipts_spec.js
Comment (Jaguar0625) [reply catapult-sdk/test/plugins/receipts_spec.js:64]: either is fine

File: rest/src/plugins/receipts/ReceiptsDb.js
Comment (fboucquez) [reply rest/src/plugins/receipts/ReceiptsDb.js:136]: Thanks for the explanation @Jaguar0625 . 

I considered the index map but my thinking was, the max page size is 100 (default is 20), in general, the worst case would be 100 individual blocks. The height field is a Long object, I cannot index by height as-is. I would need to create a Map<string, Block> where the index/key is a string representation of the Long (probably decimal of hex representation). The cost of converting the heights to string (to create the map and when looking by height index) could be worse than just the O(nlog(n)) of 100 elements, plus the memory cost of the temp map.

Do you have a better way of doing the index map? I can change it if you like.

File: rest/test/db/CatapultDb_spec.js
Comment (fboucquez) [reply rest/test/db/CatapultDb_spec.js:446]: Uff, missed that. I'll add

File: rest/src/plugins/receipts/ReceiptsDb.js
Comment (fboucquez) [reply rest/src/plugins/receipts/ReceiptsDb.js:146]: Although it's not expected the statements to have meta right now, I've added the original meta + a comment clarifying this.

File: catapult-sdk/test/plugins/receipts_spec.js
Comment (fboucquez) [reply catapult-sdk/test/plugins/receipts_spec.js:64]: https://github.com/symbol/symbol-openapi/issues/279 for the open api change

File: rest/src/plugins/receipts/ReceiptsDb.js
Comment (Jaguar0625) [reply rest/src/plugins/receipts/ReceiptsDb.js:138]: up to @gimre-xymcity

**Review by Jaguar0625 — APPROVED** [review]

File: rest/src/plugins/receipts/ReceiptsDb.js
Comment (fboucquez) [reply rest/src/plugins/receipts/ReceiptsDb.js:138]: @gimre-xymcity , are happy about the answer? Do you want the list to be indexed?

Comment (gimre-xymcity) [reply rest/src/plugins/receipts/ReceiptsDb.js:138]: don't mind

**Pull request merged**
